policybot/pkg/testresults/testresults.go

```diff
@@ -0,0 +1,378 @@
+ * Use `prFlakeyTest, err := NewPrFlakeTest()` and `testFlakes := prFlakeyTest.checkTestFlakesForPr(<pr_number>)`
+ * to get test flakes information for a given pr.
+ */
+func NewPrResultTester(bucketName string) (*PrResultTester, error) {
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &PrResultTester{
+		Client:     client,
+		ctx:        ctx,
+		bucketName: bucketName,
+	}, nil
+}
+
+/*
+ * Contains function check if a string exists in a given slice of strings.
+ */
+func contains(slic []string, ele string) bool {
+	for _, e := range slic {
+		if strings.Compare(e, ele) == 0 {
+			return true
+		}
+	}
+	return false
+}
+
+/*
+ * GetTest function get all directories under the given pr in istio-prow/pr-logs/pull/istio-istio/PRNUMBER for each test suite name.
+ * Client: client used to get buckets and objects.
+ * PrNum: the PR number inputted.
+ * Return []Tests return a slice of Tests objects.
+ */
+func (prt *PrResultTester) getTests(prNumInt int64) ([]tests, error) {
+	prNum := strconv.FormatInt(prNumInt, 10)
+	ctx := prt.ctx
+	client := prt.Client
+	bucket := client.Bucket(prt.bucketName)
+	query := &storage.Query{Prefix: "pr-logs/pull/istio_istio/" + prNum}
+	it := bucket.Objects(ctx, query)
+	var testSlice []tests
+	for {
+		attrs, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		name := attrs.Name
+
+		nameSlice := strings.Split(name, "/")
+		prNum = nameSlice[3]
+		pullNum := nameSlice[5]
+		testName := nameSlice[len(nameSlice)-3]
+		fileName := nameSlice[len(nameSlice)-1] // C
+		var newString = "pr-logs/pull/istio_istio/" + prNum + "/" + testName + "/" + pullNum
+		if strings.Compare(fileName, "started.json") == 0 || strings.Compare(fileName, "clone-records.json") == 0 || strings.Compare(fileName, "finished.json") == 0 {
+			var contain = false
+			for ind, ele := range testSlice {
+				if strings.Compare(ele.Name, testName) == 0 {
+					prs := ele.Prs
+
+					if !contains(prs, newString) {
+						prs = append(prs, newString)
+						ele.Prs = prs
+						testSlice[ind] = ele
+					}
+
+					contain = true
+				}
+
+			}
+			if !contain {
+				t := tests{
+					Name: testName,
+				}
+				newSlice := []string{newString}
+				t.Prs = newSlice
+				testSlice = append(testSlice, t)
+			}
+		}
+	}
+	return testSlice, nil
+}
+
+func (prt *PrResultTester) getInformationFromFinishedFile(pref string, onePull *store.TestResult) (*store.TestResult, error) {
+	// It is possible that the folder might not contain finished.json.
+	client := prt.Client
+	bucket := client.Bucket(prt.bucketName)
+	newObj := bucket.Object(pref + "/finished.json")
+	nrdr, nerr := newObj.NewReader(prt.ctx)
+	if nerr != nil {
+		return nil, nerr
+	}
+
+	defer nrdr.Close()
+	slur, err := ioutil.ReadAll(nrdr)
+	if err != nil {
+		return nil, nerr
+	}
+	ns := string(slur)
+	ndec := json.NewDecoder(strings.NewReader(ns))
```

Comment:
It seems all of your json decoding logic could be replaced concisely by https://godoc.org/encoding/json#Unmarshal